### DISCUSSION TITLE
Manipulating AST node for code generation

### DISCUSSION THREAD

**Discussion by luan-xiaokun** [eclipse-langium/langium#800] [Q&A] [answered]
Hey guys, I'm trying to implement a code generator for my DSL. Automatons are the fist-class citizens in this language, and they can be nested, for example, you can instantiate one automaton in another one, so I need to unfold all automatons into only one automaton, and then transform it into some target language. (This is very like unfolding every inline functions to get a big function, but it is a bit more complicated).

Therefore, I need to manipulate AST nodes, creating new ones and relate them together. For example, in some cases, I need to take two expressions from two statements, and create a disjunction of them. Then I need to write something like
```ts
const disjunct: BinaryExpression = {
    $container: container,
    $type: 'BinaryExpression'
    operator: '||',
    left: expr1,
    right: expr2
}
```
There are two things bothering me:
1. As both `expr1` and `expr2` are taken from the parsed AST, their containers are not the node `disjunct` that I'm creating, so I need to clone them and at the same time change their `$container` properties, and this process should be recursive.
2. Creating `disjunct` requires a container, and in many cases, I'm creating `disjunct` because I need it in a larger container being constructed. We cannot first create `disjunct` and then use it to construct its constructor, and finally reset the container of `disjunct`, because `$container` is a readonly property.

Given enough time, I can find workarounds for these problems, but the progress is very slow because I'm not familiar with TypeScript (nor JavaScript). Do you have any advice on dealing with these inconveniences? Both suggestions on how to do code generation and tips on useful TypeScript features are appreciated.

By the way, I have a prototype of the generator written in Java, it looks like
```java
tnew.setGuard(new BinaryOperatorTerm()
                        .setParent(tnew)
                        .setOpr(EnumBinaryOperator.LAND)
                        .setLeft(
                                new SingleOperatorTerm()
                                        .setOpr(EnumSingleOperator.LNOT)
                                        .setTerm(cond.copy(parent))
                        )
                        .setRight(tnew.getGuard())
                );
```
which is more intuitive to me (although I'm not familiar with Java either).

**Comment by msujew** [comment, marked as answer]
Hey @luan-xiaokun,

for cases such as these, we do have a `linkContentToContainer` function in `langium`, which automatically sets the `$container` property on all its direct descendants (not it's transitive descendants!). You can ignore all the `$container/$containerProperty/$containerIndex` properties during construction and the mentioned function will take care of it.

**Reply by luan-xiaokun** [reply]
That's a handy utility function, I was using `const newNode = Object.assign(node)` to change the readonly property.

Is there any way to use something like `undefined` to first create the node and then call `linkContentToContainer`? If I don't provide a container with the correct type, TypeScript will complain about the missing property. (I miss python's dynamic types）

**Reply by msujew** [reply]
Yeah, you can just use `$container: undefined!` which basically tricks typescript into thinking you've assigned something.

**Reply by luan-xiaokun** [reply]
Wow, that's a new trick for me, thanks for the help!

**Reply by spoenemann** [reply]
Nicer alternative:
```typescript
type Programmatic<T extends AstNode> = Omit<T, '$container'>
```
